DxDispatch/src/dxdispatch/CommandLineArgs.cpp

```diff
@@ -544,4 +554,36 @@ void CommandLineArgs::SetAdapter(IAdapter* adapter)
             THROW_HR(E_NOTIMPL);
         }
     }
+}
+
+DML_FEATURE_LEVEL GetDmlFeatureLevelFromString(const std::string& featureLevel)
+{
+    if ("1.0" == featureLevel)
+        return DML_FEATURE_LEVEL_1_0;
+    else if ("2.0" == featureLevel)
+        return DML_FEATURE_LEVEL_2_0;
+    else if ("2.1" == featureLevel)
+        return DML_FEATURE_LEVEL_2_1;
+    else if ("3.0" == featureLevel)        
+        return DML_FEATURE_LEVEL_3_0;
+    else if ( "3.0" == featureLevel)
+        return DML_FEATURE_LEVEL_3_1;
+    else if ("4.0" == featureLevel)
+        return DML_FEATURE_LEVEL_4_0;
+    else if ("4.1" == featureLevel)
+        return DML_FEATURE_LEVEL_4_1;
+    else if ("5.0" == featureLevel)
+        return DML_FEATURE_LEVEL_5_0;
+    else if ("5.1" == featureLevel)
+        return DML_FEATURE_LEVEL_5_1;
+    else if ("5.2" == featureLevel)
+        return DML_FEATURE_LEVEL_5_2;
+    else if ("6.0" == featureLevel)
+        return DML_FEATURE_LEVEL_6_0;
+    else if ("6.1" == featureLevel)
+        return DML_FEATURE_LEVEL_6_1;
+    else if ("6.2" == featureLevel)
+        return DML_FEATURE_LEVEL_6_2;
```

Comment:
Hmm, this else-if chain will be tedious to add to, and do we actually *need* to restrict them to a specific list? We could release (unlikely, but could) an intermediate 5.3 patch in the future. Also you *can* ask CreateDevice1 for a version like 2.2 which was never released, and the API will do the right thing. There are even tests for this in DmlDeviceTestsImpl.cpp:

```c++
    // Hypothetical future scenario where caller asks for a *lower* feature level we've never heard of. This should
    // succeed even though we've never heard of this feature level (it's not a valid enum value) because feature levels
    // are always supersets.
    static constexpr DML_FEATURE_LEVEL c_lowerFeatureLevels[] =
    {
        static_cast<DML_FEATURE_LEVEL>(0),
        static_cast<DML_FEATURE_LEVEL>(1),
        static_cast<DML_FEATURE_LEVEL>(42),
        static_cast<DML_FEATURE_LEVEL>(-1), // To ensure DML compares these as signed
    };
```

So, it could just be...

```c++
float floatFeatureLevel = atof(featureLevel);
uint32_t majorFeatureLevel = static_cast<uint32_t>(floatFeatureLevel);
uint32_t minorFeatureLevel = static_cast<uint32_t>(floatFeatureLevel - majorFeatureLevel);
return static_cast<DML_FEATURE_LEVEL>(majorFeatureLevel << 12 | minorFeatureLevel << 8);
```

...and DML will do the validation for us, without DxDispatch repeating its own validation logic.

(then the brain-bending `if` statements that appreciate Yoda would can fade away too :b)

Reply:
Also note that if you ask for DML FL 5.0, you still can still create 6.2 operators and QI for 6.2 objects. It just sets a *minimum* baseline up-front in the creation request.

Reply:
I was told to update the version since we deploy DML and ONNX runtime can depend on latest version

Reply:
> I was told to update the version since we deploy DML and ONNX runtime can depend on latest version

Alrighty. Well deferring to Justin about the update, but it's nice for DxDispatch to not *require* a newer version of DML than DxDispatch actually supports itself (meaning if DxDispatch hasn't been updated for a few versions of DML and lacks 6.2 operators, then DxDispatch would preferentially only require whatever the last version of DML was which DxDispatch was updated for).

The other aspect is the constraining version list, which we probably don't need, since DML already does this feature level validation.

Reply:
The ideal feature level to pass in would one decided by one of the following:
1. If dxdispatch is running an ONNX dispatchable: the minimum feature level required by the version of onnxruntime.dll that is loaded into the process.
2. If dxdispatch is running a DML dispatchable: the minimum feature level of the operators in the JSON model.

Without writing a bunch more code to understand what's required in 1/2 above, I'm not sure that it makes a ton of sense to specify a minimum feature level that is always the highest value (6.2 currently). Using the latest feature level just forces DML device creation to fail unnecessarily even if it can support the ONNX/DML dispatchable.